File: tools/generate_chapters.py
#!/usr/bin/env python3
import sys, os, json, re
def extract_title(text, fallback):
    m = re.search(r"<title>([\s\S]*?)</title>", text, re.I)
    return (m.group(1).strip() if m else None) or fallback
if len(sys.argv) < 2:
    print("Usage: python tools/generate_chapters.py <book_folder>"); sys.exit(1)
directory = sys.argv[1]
files = sorted([f for f in os.listdir(directory) if f.lower().endswith(".xhtml")], key=lambda s: [int(t) if t.isdigit() else t for t in re.split(r'(\d+)', s)])
chapters = []
for f in files:
    p = os.path.join(directory, f)
    with open(p, "r", encoding="utf-8") as fh:
        txt = fh.read()
    title = extract_title(txt, os.path.splitext(f)[0])
    chapters.append({"file": f, "title": title})
out = os.path.join(directory, "chapters.json")
with open(out, "w", encoding="utf-8") as fh:
    json.dump({"chapters": chapters}, fh, indent=2)
print(f"Wrote {len(chapters)} chapters to {out}")

File: tools/generate-chapters.js
// Node.js generator
const fs=require('fs');const path=require('path');
function extractTitle(text,fallback){const m=text.match(/<title>([\s\S]*?)<\/title>/i);return (m?m[1].trim():null)||fallback;}
const dir=process.argv[2]; if(!dir){console.error("Usage: node tools/generate-chapters.js <book_folder>");process.exit(1);}
const abs=path.resolve(dir);
const files=fs.readdirSync(abs).filter(f=>f.toLowerCase().endsWith('.xhtml')).sort((a,b)=>a.localeCompare(b,undefined,{numeric:true}));
const chapters=files.map(f=>{const txt=fs.readFileSync(path.join(abs,f),'utf8');return {file:f,title:extractTitle(txt, path.parse(f).name)};});
fs.writeFileSync(path.join(abs,'chapters.json'), JSON.stringify({chapters}, null, 2));
console.log(`Wrote ${chapters.length} chapters to ${path.join(abs,'chapters.json')}`);
